### PR TITLE
Instrumentation via cowboy_telemetry

### LINKED ISSUE #42: Adopt cowboy_telemetry
We are creating a new Cowboy stream handler that will emit telemetry events:

* https://github.com/beam-telemetry/cowboy_telemetry

Once it's released, we should adopt it in `plug_cowboy` and include it by default, and also adopt it for the `early_error` logging.

### PR DESCRIPTION
**Pull request by binaryseed**
This PR pulls in the new `cowboy_telemetry` instrumentation library to provide safe instrumentation that is not susceptible to the problems that were inherent in the previous `plug_adapter` instrumentation approach.

* `Plug.Cowboy.Stream` handler was replaced by the default `:cowboy_stream_h`
* Plug's existing `413` response logging is now done by listening for `[:cowboy, :request, :early_error]` events
* Tests validate the basic instrumentation events that are provided

closes #42

### CODE REVIEW

**Comment by binaryseed** [comment]
@josevalim I would like a review on this..

I'm curious what you think regarding the best way to roll this out... I'm working on putting together and deploying a test app that leverages the `cowboy_telemetry_h` instrumentation on it's own via my agent, so that should help validate the basics...

**Comment by josevalim** [comment]
Sorry for the delay, it looks absolutely great! My proposal is to merge this and then you give it a try by using a git dependency. :)

**Comment by binaryseed** [comment]
I think I'd like us to consider the approach in https://github.com/beam-telemetry/cowboy_telemetry/pull/6 before we go forward with merging this into `plug_cowboy` - it's far simpler since it delegates to cowboy's own `cowboy_metrics_h`, but some of the metadata changes...

**Comment by binaryseed** [comment]
@josevalim I'm getting back to this, and I'd like to rebase on the latest release, but it looks like `v2.3.0` is a tag only, and isn't yet part of `master`... Perhaps this  release commit (https://github.com/elixir-plug/plug_cowboy/commit/2fb3285ea8f0e5302a5ab2b62abf9bab797c9b98) needs to be pulled into `master`?

**Comment by binaryseed** [comment]
Also the CI build is failing on compiling cowboy. I see that `plug` was migrated over to Github Actions, perhaps we should do that here too...

**Comment by josevalim** [comment]
Updated master. The issue with CI is that Cowboy doesn't work on more recent OTP versions. So if you can stay within 2.7.0, you should be fine. You can for example temporarily update mix.exs to depend on ~> 2.7.0, do deps.get, and then revert the mix.exs change. Also pinging @adrianomitre in case he wants to send a PR to update CI. :heart:

**Comment by binaryseed** [comment]
I rebased & updated this, but it looks like we're still seeing the `cowboy` compilation failures :(

File: mix.lock [hunk 1]
```diff
@@ -1,6 +1,7 @@
 %{
-  "cowboy": {:hex, :cowboy, "2.7.0", "91ed100138a764355f43316b1d23d7ff6bdb0de4ea618cb5d8677c93a7a2f115", [:rebar3], [{:cowlib, "~> 2.8.0", [hex: :cowlib, repo: "hexpm", optional: false]}, {:ranch, "~> 1.7.1", [hex: :ranch, repo: "hexpm", optional: false]}], "hexpm", "04fd8c6a39edc6aaa9c26123009200fc61f92a3a94f3178c527b70b767c6e605"},
-  "cowlib": {:hex, :cowlib, "2.8.0", "fd0ff1787db84ac415b8211573e9a30a3ebe71b5cbff7f720089972b2319c8a4", [:rebar3], [], "hexpm", "79f954a7021b302186a950a32869dbc185523d99d3e44ce430cd1f3289f41ed4"},
+  "cowboy": {:hex, :cowboy, "2.8.0", "f3dc62e35797ecd9ac1b50db74611193c29815401e53bac9a5c0577bd7bc667d", [:rebar3], [{:cowlib, "~> 2.9.1", [hex: :cowlib, repo: "hexpm", optional: false]}, {:ranch, "~> 1.7.1", [hex: :ranch, repo: "hexpm", optional: false]}], "hexpm", "4643e4fba74ac96d4d152c75803de6fad0b3fa5df354c71afdd6cbeeb15fac8a"},
```
Comment (josevalim) [inline mix.lock:2]: Can we depend on cowboy 2.7.0 still? or does cowboy_telemetry require later versions?

Comment (binaryseed) [reply mix.lock]: Yeah, that was only updated as a side effect, there isn't a hard requirement on `2.8`

Comment (josevalim) [reply mix.lock]: I was more hoping that it would fix the compilation issues. :)

Comment (binaryseed) [reply mix.lock]: it did fix the compilation issues, but the `(1.10.4, 23.0.3` tests are still very flaky

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**